### ISSUE TITLE
CVE-2018-10237

### ISSUE THREAD

**Issue by patrickhousley** [Adobe-Consulting-Services/acs-aem-commons#1619]
## Required Information

* [X] AEM Version, including Service Packs, Cumulative Fix Packs, etc: 6.3.3.0
* [X] ACS AEM Commons Version: 3.17.0
* [X] Reproducible on Latest? yes

### Expected Behavior
Code should pass vulnerability scan.

### Actual Behavior
Code fails vulnerability scan with below output.

```bash
One or more dependencies were identified with known vulnerabilities in common AEM Project - Reactor Project:

guava-19.0.jar (cpe:/a:google:guava:19.0, com.google.guava:guava:19.0) : CVE-2018-10237
acs-aem-commons-bundle-3.17.0.jar: guava-15.0.jar (cpe:/a:google:guava:15.0, com.google.guava:guava:15.0) : CVE-2018-10237
```

Latest version:
```bash
guava-19.0.jar (cpe:/a:google:guava:19.0, com.google.guava:guava:19.0) : CVE-2018-10237
acs-aem-commons-bundle-3.19.0.jar (cpe:/a:adobe:dispatcher:3.19.0, cpe:/a:apache:commons_email:3.19.0, cpe:/a:apache:commons-httpclient:3.19.0, com.adobe.acs:acs-aem-commons-bundle:3.19.0) : CVE-2016-0957
acs-aem-commons-bundle-3.19.0.jar: guava-15.0.jar (cpe:/a:google:guava:15.0, com.google.guava:guava:15.0) : CVE-2018-10237
```

### Steps to Reproduce
Implement `org.owasp.dependency-check-maven` in this code repository and perform an analysis. Repository should fail with violation `CVE-2018-10237`.

### Links
https://github.com/google/guava/wiki/CVE-2018-10237
https://www.cvedetails.com/cve/CVE-2018-10237/

**Comment by BillDinger** [comment]
`dependency-check-maven`  is using the [owasp dependency checker](https://www.owasp.org/index.php/OWASP_Dependency_Check) but example of pom implementation:

```
        <plugin>
          <groupId>org.owasp</groupId>
          <artifactId>dependency-check-maven</artifactId>
          <version>3.1.2</version>
          <configuration>
            <cveValidForHours>24</cveValidForHours>
            <failBuildOnCVSS>4</failBuildOnCVSS>
          </configuration>
          <executions>
            <execution>
              <goals>
                <goal>check</goal>
              </goals>
            </execution>
          </executions>
        </plugin>
```

with the following command to execute : `mvn dependency-check:check`

**Comment by BillDinger** [comment]
so I just added this to the project [in my fork](https://github.com/BillDinger/acs-aem-commons/commit/2ab1448aa6756b5dcc7e155f50bd23611ab50d5b) and the current list of vulnerabilities is:

```
[ERROR]
[ERROR] One or more dependencies were identified with vulnerabilities that have a CVSS score greater than or equal to '4.0':
[ERROR]
[ERROR] jcr-2.0.jar: CVE-2017-16111
[ERROR] commons-email-1.2.jar: CVE-2017-9801, CVE-2018-1294
[ERROR] logback-core-1.0.13.jar: CVE-2017-5929
[ERROR] com.adobe.granite.poi-2.0.6.jar: CVE-2014-9527, CVE-2016-5000, CVE-2017-5644, CVE-2014-3574, CVE-2012-0213, CVE-2014-3529
[ERROR] tika-core-1.14.jar: CVE-2018-1335, CVE-2018-8017, CVE-2018-1338, CVE-2018-11761, CVE-2018-11762, CVE-2018-1339
[ERROR] tika-parsers-1.5.jar: CVE-2016-6809, CVE-2018-1335, CVE-2018-8017, CVE-2018-1338, CVE-2018-11761, CVE-2018-11762, CVE-2018-1339
[ERROR] vorbis-java-tika-0.1.jar: CVE-2016-6809, CVE-2018-1335, CVE-2017-6888, CVE-2018-1338, CVE-2018-11761, CVE-2018-1339
[ERROR] xz-1.2.jar: CVE-2015-4035
[ERROR] pdfbox-1.8.4.jar: CVE-2018-8036, CVE-2016-2175
[ERROR] fontbox-1.8.4.jar: CVE-2018-8036, CVE-2016-2175
[ERROR] jempbox-1.8.4.jar: CVE-2018-8036, CVE-2016-2175
[ERROR] bcprov-jdk15-1.45.jar: CVE-2016-1000341, CVE-2016-1000352, CVE-2013-1624, CVE-2015-7940, CVE-2018-1000613, CVE-2016-1000339, CVE-2016-1000346, CVE-2016-1000338, CVE-2017-13098, CVE-2016-1000343, CVE-2016-1000342, CVE-2018-5382, CVE-2016-1000345, CVE-2016-1000344
[ERROR] poi-3.10-beta2.jar: CVE-2014-9527, CVE-2016-5000, CVE-2017-5644, CVE-2014-3574, CVE-2014-3529, CVE-2017-12626
[ERROR] dom4j-1.6.1.jar: CVE-2018-1000632
[ERROR] org.apache.sling.models.api-1.3.0.jar: CVE-2013-4390, CVE-2016-0956, CVE-2015-2944
[ERROR] jjwt-0.7.0.jar: CVE-2018-7308
[ERROR] org.apache.sling.servlets.annotations-1.0.0.jar: CVE-2016-5394, CVE-2013-4390, CVE-2016-0956, CVE-2016-6798
[ERROR] org.apache.sling.api-2.16.0.jar: CVE-2013-4390, CVE-2016-0956
[ERROR] aws-java-sdk-osgi-1.10.76.jar\META-INF/maven/com.amazonaws/aws-java-sdk-core/pom.xml: CVE-2006-1643, CVE-2007-4177, CVE-2006-1644
[ERROR]
```

Some of these might be test only.

**Comment by justinedelson** [comment]
Unfortunately, I don't think this can be resolved. ACS AEM Commons depends upon Oak for a number of tests and Oak is still using 15.0 and, specifically using a method which has since been made package-private (`com.google.common.collect.Iterators.emptyIterator()`).

@BillDinger that list is including provided scope dependencies (which is the default for that plugin). The only relevant vulnerabilities IMO are those in embedded dependencies. Your list does include jjwt which is an embedded dependency, but that CVE is related to some PHP application, so I'm confused by that one.

**Comment by justinedelson** [comment]
In #1620, I added the dependency check plugin, although had to exclude guava for the reasons stated above. The plugin is configured to only check embedded dependencies (currently guava, jsch, jjwt, and jzlib).

**Comment by BillDinger** [comment]
thanks @justinedelson you beat me to the PR . I think there's an argument for changing to scoped dependencies so we try and force the correct, non vulnerable dependencies to be provided

**Comment by justinedelson** [comment]
I'm not sure what you mean by "scoped dependencies" but dependency versions cannot be changed at will -- most of those dependencies are provided by the base platform and *must* be the exact versions used by the minimal compatible version of the platform. This project should not recommend users do something which would impact their ability to obtain support from Adobe.

That said, I suspect that some of these dependencies may be outdated and as part of changing the minimal version from 6.2 to 6.3, some cleanup would be useful. For example, I'm not sure that the explicit Sling Model API dependency is necessary if the minimal AEM version is 6.3.

**Comment by BillDinger** [comment]
Meant provided not scoped apologies. Changing to 6.3 provided dependencies if that is the minimum we are now supporting would clear up most of these. Went through similar when cleaning up our own provided dependencies when going from 6.1 -> 6.3 and virtually everything had been updated or was updated in subsequent fix packs

**Comment by justinedelson** [comment]
ah, yeah. there's actually a configuration to ban *all* compile-scope dependencies by default. I'd personally recommend every AEM project do that :)

**Comment by justinedelson** [comment]
one bit of follow up - I confirmed with some internal documentation that AEM is *not* vulnerable to this particular issue because of the deserialization firewall.

**Comment by justinedelson** [comment]
...but because suppression is lame (and Oak 1.8.4+ gets rid of the incompatibility - https://issues.apache.org/jira/browse/OAK-7186), I went ahead and updated Guava anyway.

**Comment by justinedelson** [comment]
fixed by #1621
